### PR TITLE
[labs/react] Provide explicit return type from createComponent

### PR DESCRIPTION
**Pull request by taylor-vann**
Related to:
https://github.com/lit/lit/issues/3188

After landing, revisit:
https://github.com/lit/lit/issues/3021
https://github.com/lit/lit/issues/2951

This PR is a type-only change (except for one location described below).

This PR exports an explicit return type `ReactWebComponent` for Typescript for downstream users. This allows users to prepare for Types without calling the `createComponent` function. An explicit return type also enables overrides for `createComponent` arguments.

A successor PR https://github.com/lit/lit/pull/2988 will create `params` object override.

++ The only change outside of types is a `null` used to diff check `refs`. However even this change is made to satisfy typing constraints on `refs`. So it's _almost_ a type change.

```ts
246      const userRef = this.props.__forwardedRef ?? null;
```

### CODE REVIEW

**Review by sorvell — CHANGES_REQUESTED** [review]
Please change the description of this PR to first clearly state the problem with links to the provoking issue(s). Then discuss the changes. We also need to ensure there are tests to validate that the changes fix the referenced issues. Seems doable at least for https://github.com/lit/lit/issues/3021.

Thanks.

**Comment by taylor-vann** [comment]
@sorvell updated the title to fit the PR better!

As far as providing tests to fix related issues, I feel that's a little in depth and out of scope for this PR.

I don't think this one typing change would fix an issue like https://github.com/lit/lit/issues/2951. But it would eliminate some of the ambiguity described in https://github.com/lit/lit/issues/2951.

The purpose of this PR is to harden types in the react package so we can eliminate the wrapper as a source of type ambiguity.

**Comment by taylor-vann** [comment]
I've updated the description and reduced the PR to the most minimal change.
This is ready for a review review.

File: packages/labs/react/src/create-component.ts [hunk 1]
```diff
@@ -4,6 +4,47 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
+// Typecast that curries an Event type through a string to match
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:7]: Can we remove the "curries" part here? It's not consistent with my understanding of the definition of the term.

File: packages/labs/react/src/create-component.ts [hunk 2]
```diff
@@ -4,6 +4,47 @@
...
+// a prop name to a typed event callback.
+export type EventName<T extends Event = Event> = string & {
+  __event_type: T;
+};
+
+type Events = Record<string, EventName | string>;
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:13]: For our own understanding I'd like to comment this and `EventProps`. I keep forgetting why we have both.

I'd also like to consider another name for `Events`. Is it an `EventMap` maybe?

File: packages/labs/react/src/create-component.ts [hunk 3]
```diff
@@ -4,6 +4,47 @@
...
+    ? (e: R[K]['__event_type']) => void
+    : (e: Event) => void;
+};
+
+type ReactProps<I, E> = Omit<React.HTMLAttributes<I>, keyof E>;
+type ElementWithoutPropsOrEvents<I, E> = Omit<
+  I,
+  keyof E | keyof ReactProps<I, E>
+>;
+// Props the user is allowed to use, includes standard attributes, children,
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:26]: style: add newline

File: packages/labs/react/src/create-component.ts [hunk 4]
```diff
@@ -4,6 +4,47 @@
...
+// ref, as well as special event and element properties.
+export type ElementProps<I, E extends Events = {}> = Partial<
+  ReactProps<I, E> & ElementWithoutPropsOrEvents<I, E> & EventProps<E>
+>;
+// Props used by this component wrapper. This is the ElementProps and the
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:31]: style: add newline

File: packages/labs/react/src/create-component.ts [hunk 5]
```diff
@@ -4,6 +4,47 @@
...
+// special `__forwardedRef` property. Note, this ref is special because
+// it's both needed in this component to get access to the rendered element
+// and must fulfill any ref passed by the user.
+type ComponentProps<I, E extends Events = {}> = ElementProps<I, E> & {
+  __forwardedRef?: React.Ref<I>;
+};
+
+export type WrappedWebComponent<
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:39]: How about `ReactWebComponent`?

File: packages/labs/react/src/create-component.ts [hunk 6]
```diff
@@ -255,17 +264,19 @@ export const createComponent = <I extends HTMLElement, E extends Events = {}>(
...
   }
 
-  const ForwardedComponent = React.forwardRef(
-    (props?: UserProps, ref?: React.Ref<unknown>) =>
-      createElement(
-        ReactComponent,
-        {...props, __forwardedRef: ref} as ComponentProps,
-        props?.children
-      )
+  const ForwardedComponent: WrappedWebComponent<I, E> = React.forwardRef<
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:271]: Can you add the return type to `createComponent` on line 153?

**Review by justinfagnani — CHANGES_REQUESTED** [review]

File: packages/labs/react/src/create-component.ts (hunk 2 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:13]: How about `EventNameMap` since we also use `EventName` as that type?

File: packages/labs/react/src/create-component.ts (hunk 1 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:7]: sounds good, curry is farther than where i started, makes sense

File: packages/labs/react/src/create-component.ts (hunk 2 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:13]: `EventName` -> `EventNames` -> `EventNameListeners` ?

File: packages/labs/react/src/create-component.ts (hunk 6 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts]: done!

File: packages/labs/react/src/create-component.ts (hunk 5 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts]: sounds good to me!

File: packages/labs/react/src/create-component.ts (hunk 4 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:40]: done

File: packages/labs/react/src/create-component.ts (hunk 3 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:29]: done

File: packages/labs/react/src/create-component.ts (hunk 1 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts]: done!

**Comment by taylor-vann** [comment]
Tests are passing.

This PR is no longer blocked and ready for submission.

File: packages/labs/gen-wrapper-react/package.json [hunk 7]
```diff
@@ -60,11 +60,13 @@
...
   },
   "dependencies": {
+    "@lit-internal/tests": "0.0.0",
     "@lit-labs/analyzer": "^0.2.0",
     "@lit-labs/gen-utils": "^0.1.0"
   },
   "devDependencies": {
-    "@lit-internal/tests": "^0.0.0"
+    "@lit-internal/tests": "^0.0.0",
+    "@types/react": "^18.0.17"
```
Comment (kevinpschaaf) [inline packages/labs/gen-wrapper-react/package.json:69]: I understand that this fixed the type compilation issue, but I don't understand why. The `gen-wrapper-react` source does not actually do anything with react (note, it doesn't have react as a dependency), it only generates and tests code that uses react.

The generated react wrapper has a peer dependency on `@types/react` (see [generator](https://github.com/lit/lit/blob/main/packages/labs/gen-wrapper-react/src/index.ts#L97) and [golden](https://github.com/lit/lit/blob/main/packages/labs/gen-wrapper-react/goldens/test-element-a/package.json#L15)), and the test folder into which the generated wrapper is installed depends on `@types/react` [[source](https://github.com/lit/lit/blob/main/packages/labs/gen-wrapper-react/test-output/package.json#L10)].

So it's unclear in my mind what is resolving `@types/react` down into `gen-wrapper-react`'s node_modules.  I want to understand that, otherwise it's not clear we're missing something that will hit actual users, since the direct user of the generate wrapper having a dep on `@types/react` apparently isn't sufficient.

**Review by sorvell — CHANGES_REQUESTED** [review]
I'd like to see a test that leverages and concretely demonstrates the value of this change described in the PR description.

File: packages/labs/gen-wrapper-react/package.json (hunk 7 above)
Comment (taylor-vann) [reply packages/labs/gen-wrapper-react/package.json]: Oh nice, i think you just unlocked something for me!

in the lit-labs/react pacakge.json there's a `types` property
```
{
...
    "types": ["react", "react-dom", "mocha"]
}
```

and when i add that to the `lit-labs/gen-wrapper-react/test-output/package.json`
```
{
...
    "types": ["react", "react-dom", "mocha"]
}
```

We get passing tests for `gen-wrapper-react` without adding `@types/react` to the gen-wrapper-react package.

Comment (justinfagnani) [reply packages/labs/gen-wrapper-react/package.json]: That confuses me. The `"types"` field is supposed to limit what `@types/` packages are included, not expand them: https://www.typescriptlang.org/tsconfig#types

Comment (taylor-vann) [reply packages/labs/gen-wrapper-react/package.json]: even more confusing it fails windows tests but passes others

Comment (taylor-vann) [reply packages/labs/gen-wrapper-react/package.json]: without the `types` array, all local tests fail on unix and windows

Comment (taylor-vann) [reply packages/labs/gen-wrapper-react/package.json]: I'm trying to discern whether we are operating within the domain of our npm workspace or if the test-output is completely separate from the workspace.

Adding the @types resolves this but AFAIU @kevinpschaaf we don't want the @types associated with `gen-wrapper-react` only the `test-output`.

This feels in the domain of ownership and npm workspaces. It makes me think that `gen-wrapper-react` is the domain and `test-output` is operating within that domain.

The `test-output` is independent and sort of an integration test for the wrapper no? But do we really know how isolated `test-output` from `gen-wrapper-react`? That's currently where I'm at trying to rationalize why adding `@types` to `gen-wrapper-react` works here.

It would make me wonder if `test-output` was entirely outside the Lit package (a real world use case) would resolve this.

Comment (taylor-vann) [reply packages/labs/gen-wrapper-react/package.json]: is it possible that we need to make `test-output` a workspace?

File: packages/labs/react/src/test/create-component_test.tsx [hunk 8]
```diff
@@ -131,6 +131,18 @@ suite('createComponent', () => {
...
+  /*
+    The following test is a type-only test.
+  */
+  test('renders element with expected type', async () => {
+    type TypedComponent = ReactWebComponent<BasicElement>;
+
+    let TypedBasicElement!: TypedComponent;
+
+    // @ts-expect-error
+    <TypedBasicElement bool={"string"}></TypedBasicElement>
```
Comment (sorvell) [inline packages/labs/react/src/test/create-component_test.tsx:143]: Add a comment explaining why this should be a type error.

File: packages/labs/react/src/create-component.ts [hunk 9]
```diff
@@ -132,31 +170,39 @@ export const createComponent = <I extends HTMLElement, E extends Events = {}>(
...
-  // Props used by this component wrapper. This is the UserProps and the
-  // special `__forwardedRef` property. Note, this ref is special because
-  // it's both needed in this component to get access to the rendered element
-  // and must fulfill any ref passed by the user.
-  type ComponentProps = UserProps & {
-    __forwardedRef?: React.Ref<unknown>;
-  };
+  // Set of properties/events which should be specially handled by the wrapper
+  // and not handled directly by React.
+  const elementClassProps = new Set(Object.keys(events ?? {}));
```
Comment (sorvell) [inline packages/labs/react/src/create-component.ts:177]: How is this changed related to the PR?

**Review by sorvell — DISMISSED** [review]

File: packages/labs/react/src/create-component.ts (hunk 9 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:177]: i'll revert

Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:177]: Oh there's nothing to revert to. This is coming across as a change but it's the same as `main`.

File: packages/labs/react/src/test/create-component_test.tsx (hunk 8 above)
Comment (taylor-vann) [reply packages/labs/react/src/test/create-component_test.tsx:147]: done!

File: packages/labs/gen-wrapper-react/package.json (hunk 7 above)
Comment (taylor-vann) [reply packages/labs/gen-wrapper-react/package.json]: resolved with https://github.com/lit/lit/pull/3310

File: packages/labs/react/src/create-component.ts (hunk 9 above)
Comment (justinfagnani) [reply packages/labs/react/src/create-component.ts:177]: I don't see this in the original file in this diff though, so it's confusing. Looks like new code. Same with the for loop below.

Do you need to merge in `main`?

Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:177]: im not trying to change this part of the code

Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:177]: well this PR has been sitting in limbo and had other submissions, not surprised there's side effects here

Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts]: removed

File: packages/labs/react/src/create-component.ts [hunk 10]
```diff
@@ -121,7 +156,10 @@ type EventProps<R extends Events> = {
  * messages. Default value is inferred from the name of custom element class
  * registered via `customElements.define`.
  */
-export const createComponent = <I extends HTMLElement, E extends Events = {}>(
+export const createComponent = <
+  I extends HTMLElement,
+  E extends EventNames = {}
+>(
```
Comment (justinfagnani) [inline packages/labs/react/src/create-component.ts:162]: We should be using `ReactWebComponent<I, E>` as the return type here now.

**Review by justinfagnani — CHANGES_REQUESTED** [review]

File: packages/labs/react/src/create-component.ts (hunk 10 above)
Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:162]: yessir!

Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:162]: lol did everything but add a return type xD

Comment (taylor-vann) [reply packages/labs/react/src/create-component.ts:162]: done!

**Review by taylor-vann — COMMENTED** [review]
changes reviewd

**Review by justinfagnani — APPROVED** [review]
LGTM!

**Review by taylor-vann — COMMENTED** [review]
changes addressed
